File: backend/app/api/schemas/backup.py
"""Pydantic schemas for the database-backup API."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel


class BackupEntryResponse(BaseModel):
    """One snapshot file as returned by GET /api/backup/list."""

    path: str
    size_bytes: int
    created_utc: datetime
    kind: Literal["daily", "pre-upgrade"]


class BackupListResponse(BaseModel):
    entries: list[BackupEntryResponse]


class BackupDirResponse(BaseModel):
    path: str


class SnapshotRequest(BaseModel):
    """If `target_dir` is set we write there; otherwise we force-write to the ring buffer."""

    target_dir: str | None = None


class SnapshotResponse(BaseModel):
    path: str
    size_bytes: int


class RestoreRequest(BaseModel):
    source_path: str


class RestoreResponse(BaseModel):
    scheduled: Literal[True] = True

File: frontend/src/api/backup.ts
/**
 * Database-backup API.
 *
 * The four endpoints under /api/backup, plus their TypeScript types.
 * Restore is two-phase: POST /restore validates and writes the
 * .restore-on-next-launch marker; the caller then asks Electron to
 * quit, and the next launch swaps the DB in before init_db runs.
 */

import { api } from "../lib/api-client";

export type BackupKind = "daily" | "pre-upgrade" | "pre-restore" | "manual";

export interface BackupEntry {
  path: string;
  size_bytes: number;
  created_utc: string;
  kind: BackupKind;
}

export interface BackupListResponse {
  entries: BackupEntry[];
}

export interface BackupDirResponse {
  path: string;
}

export interface SnapshotResponse {
  path: string;
  size_bytes: number;
}

export const backupApi = {
  getDir: () => api.get<BackupDirResponse>("/api/backup/dir"),
  list: () => api.get<BackupListResponse>("/api/backup/list"),
  snapshotToRingBuffer: () =>
    api.post<SnapshotResponse>("/api/backup/snapshot", {}),
  snapshotToFolder: (target_dir: string) =>
    api.post<SnapshotResponse>("/api/backup/snapshot", { target_dir }),
  restore: (source_path: string) =>
    api.post<{ scheduled: true }>("/api/backup/restore", { source_path }),
};
